src/main/java/com/loofah/graph/api/services/SkillService.java

```diff
@@ -18,12 +25,29 @@ public SkillService(DataRetriever dataRetriever) {
         this.dataRetriever = dataRetriever;
     }
 
-    public Skill getById(String id) {
-        return dataRetriever.getSkillById(id);
+    public SkillDTO getById(String id) {
+        Skill skill = dataRetriever.getSkillById(id).orElseThrow(() -> new DataNotFoundException("No skill found with id ["+id+"]"));
+        return transformToSkillDTO(skill);
     }
 
-    public List<Skill> getWithFilter(SkillFilter skillFilter) {
-        return dataRetriever.getSkillWithFilter(skillFilter);
+    //TODO This method will return an empty list if any of the skills that match are not retrievable (eg if the category cannot be found)
+    // Therefore in this case, an empty list will be returned to the calling client in the "data" field of the http response.
+    // This should be improved so that instead our API returns the partial list of skills that could be retrieved in the
+    // "data" field and details of those that could not be retrieved in the "errors" field of the response.
+    public List<SkillDTO> getWithFilter(SkillFilter skillFilter) {
+        List<Skill> skills = dataRetriever.getSkillWithFilter(skillFilter);
+        return skills.stream().map(this::transformToSkillDTO).collect(Collectors.toList());
+    }
+
+    private SkillDTO transformToSkillDTO(Skill skill) {
+        if (skill.getCategoryId() == null) {
+            throw new GraphQLException("Category Id is null for skill with id ["+skill.getId()+"]");
+        }
+
+        Optional<Category> category = dataRetriever.getCategoryById(skill.getCategoryId());
```

Comment:
i think at some point we talked about possibly running this second call to the category repository only if the category was requested in the incoming query using `DataFetchingFieldSelectionSet` (link with more detail on ticket). maybe we could raise an improvement ticket to do this for both category and grade at the same time once LFH-47 and LFH-49 have both gone through. let me know what you reckon :)

Reply:
oh yesss sorry this completely slipped my mind - I had intended to implement it that way.

I agree, I think an improvement ticket to do that would be perfect as the PR is already pretty large. I will raise it now

Reply:
https://devservices.jira.com/browse/LOOFAH-60 :)

Reply:
thank you :)